sway/input/keyboard.c

```diff
@@ -488,17 +503,47 @@ static void handle_key_event(struct sway_keyboard *keyboard,
 				keyinfo.raw_keysyms_len);
 	}
 
-	if (!handled || event->state == WL_KEYBOARD_KEY_STATE_RELEASED) {
-		bool pressed_sent = update_shortcut_state(
-				&keyboard->state_pressed_sent, event->keycode, event->state,
-				keyinfo.keycode, 0);
-		if (pressed_sent || event->state == WL_KEYBOARD_KEY_STATE_PRESSED) {
+	bool pressed_sent = false;
+	if (event->state == WL_KEYBOARD_KEY_STATE_RELEASED) {
+		// If the pressed event was sent to a client, also send the released
+		// event. In particular, don't send the released event to the IM grab.
+		pressed_sent = update_shortcut_state(
+			&keyboard->state_pressed_sent, event->keycode,
+			event->state, keyinfo.keycode, 0);
+		if (pressed_sent) {
 			wlr_seat_set_keyboard(wlr_seat, wlr_device);
 			wlr_seat_keyboard_notify_key(wlr_seat, event->time_msec,
-					event->keycode, event->state);
+				event->keycode, event->state);
+			handled = true;
 		}
 	}
 
+	if (!handled) {
+		struct wlr_input_method_keyboard_grab_v2 *kb_grab = keyboard_get_im_grab(keyboard);
+		struct wlr_virtual_keyboard_v1 *virtual_keyboard =
+			wlr_input_device_get_virtual_keyboard(wlr_device);
+
+		// If event is from virtual keyboard of the same client as grab,
+		// do not send it back. TODO see swaywm/wlroots#2322
+		if (kb_grab && !(virtual_keyboard &&
+				wl_resource_get_client(virtual_keyboard->resource) ==
+				wl_resource_get_client(kb_grab->resource))) {
+			wlr_input_method_keyboard_grab_v2_set_keyboard(kb_grab,
+				wlr_device->keyboard);
+			wlr_input_method_keyboard_grab_v2_send_key(kb_grab,
+				event->time_msec, event->keycode, event->state);
+			handled = true;
+		}
+	}
+
+	if (!handled && !pressed_sent) {
```

Comment:
Why do we need to check for `!pressed_sent` here?

If the event was not handled, then `pressed_sent` will always be false, or am I missing something?

> The code above would send released event to client even if no pressed event was sent before. I've fixed it.

Hm. If a pressed event was sent to the client, the first block `if (event->state == WL_KEYBOARD_KEY_STATE_RELEASED)` is responsible for sending the release event (and set `handled = true`).

If no pressed event was sent, then the first block will not handle the event, nor will the IM block. Maybe we need to restore the original condition in this block to fix it?

```c
if (pressed_sent || event->state == WL_KEYBOARD_KEY_STATE_PRESSED)
```

Reply:
I was trying to match that original check but I was wrong. I am sorry for that.

I just came up with a much simpler way to restore the original behavior. Not handling released events at that last block should do the trick. I have updated this PR for it.

Reply:
Oh, yeah, sounds good to me and simpler than my proposal.